apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +80,193 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetError(errorType string) (SubCondition, bool) {
+
+	i := getIndex(errorType, dc.Errors)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Errors[i], true
+}
+
+// AddWarning adds an warning-level Subcondition to the DetailedCondition.
+// If a SubCondition with the given warnType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddWarning(warnType, reason, message string) {
+
+	detailedReason := warnType + reason
+	dc.updateReason(detailedReason, message)
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		dc.Warnings = append(dc.Warnings, SubCondition{
+			Type:    warnType,
+			Status:  ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		})
+		return
+	}
+
+	// If the warning is already present, just update the reason and message.
+	dc.Warnings[i].updateReason(reason, message)
+}
+
+// AddWarningf adds an warning-level Subcondition to the DetailedCondition, using
+// fmt.Sprintf on the formatmsg and args params.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddWarningf(warnType, reason, formatmsg string, args ...interface{}) {
+
+	dc.AddWarning(warnType, reason, fmt.Sprintf(formatmsg, args...))
+}
+
+// GetWarning gets an warning of the given warnType.
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetWarning(warnType string) (SubCondition, bool) {
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Warnings[i], true
+}
+
+// updateReason updates a DetailedCondition's reason and message correctly
+// if they are different to the existing ones.
+func (dc *DetailedCondition) updateReason(reason, message string) {
+	if dc.Reason == "" {
+		dc.Reason = reason
+		dc.Message = message
+		return
+	}
+
+	if dc.Reason != reason {
+		dc.Reason = "MultipleProblems"
+		dc.Message = "Multiple problems were found, see errors for details"
+		return
+	}
+
+	if dc.Message != message {
+		dc.Message = dc.Message + ", " + message
+	}
+}
+
+func (sc *SubCondition) updateReason(reason, message string) {
+
+	if sc.Reason == "" {
+		sc.Reason = reason
+		sc.Message = message
+		return
+	}
+
+	if sc.Reason != reason {
+		sc.Reason = "MultipleReasons"
+	}
+
+	sc.Message = sc.Message + ", " + message
+
+}
+
+// getIndex checks if a SubCondition of type condType exists in the
+// slice, and returns its index if so. If not, returns -1.
+func getIndex(condType string, subconds []SubCondition) int {
+
+	for i, cond := range subconds {
```

Comment:
It's not guaranteed to be safe, true. I'm not sure what else to do though, with these slices that act as a map at the APIserver. I'll have a think and see if I can find something better.

Reply:
Actually, I thought about this more - I think this is probably okay, because it's operating on a given `generation` of the object. So, even if the object is updated on the APIserver, that will be picked up in another run. I think the rule here is that any particular time a thing would interact with this object should use a single version of the object, so this is an atomic operation with respect to the apiserver.